src/Juvix/LLVM/Codegen/Block.hs

```diff
@@ -0,0 +1,554 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Juvix.LLVM.Codegen.Block where
+
+import Data.ByteString.Short
+import Juvix.LLVM.Codegen.Shared
+import Juvix.LLVM.Codegen.Types as Types
+import Juvix.Library hiding (Type, local)
+import Juvix.Utility.HashMap as Map
+import LLVM.AST
+import qualified LLVM.AST as AST
+import LLVM.AST.AddrSpace
+import qualified LLVM.AST.CallingConvention as CC
+import qualified LLVM.AST.Constant as C
+import qualified LLVM.AST.Global as Global
+import qualified LLVM.AST.IntegerPredicate as IntPred
+import qualified LLVM.AST.Type as Type
+
+--------------------------------------------------------------------------------
+-- Codegen Operations
+--------------------------------------------------------------------------------
+
+fresh ∷ (HasState "count" b m, Enum b) ⇒ m b
+fresh = do
+  i ← get @"count"
+  put @"count" (succ i)
+  pure (succ i)
+
+emptyBlock ∷ Int → BlockState
+emptyBlock i = BlockState i [] Nothing
+
+createBlocks ∷ HasState "blocks" (HashMap Name BlockState) f ⇒ f [BasicBlock]
+createBlocks = fmap makeBlock . sortBlocks . Map.toList <$> get @"blocks"
+
+makeBlock ∷ (Name, BlockState) → BasicBlock
+makeBlock (l, BlockState _ s t) = BasicBlock l (reverse s) (maketerm t)
+  where
+    maketerm (Just x) = x
+    maketerm Nothing = undefined -- error $ "Block has no terminator: " <> show l
+
+sortBlocks ∷ [(a, BlockState)] → [(a, BlockState)]
+sortBlocks = sortBy (compare `on` (idx . snd))
+
+entryBlockName ∷ IsString p ⇒ p
+entryBlockName = "entry"
+
+--------------------------------------------------------------------------------
+-- Module Level
+--------------------------------------------------------------------------------
+
+emptyModule ∷ ShortByteString → Module
+emptyModule label = AST.defaultModule {moduleName = label}
+
+addDefn ∷ HasState "moduleDefinitions" [Definition] m ⇒ Definition → m ()
+addDefn d = modify @"moduleDefinitions" (<> [d])
+
+define ∷
+  HasState "moduleDefinitions" [Definition] m ⇒
+  Type →
+  Symbol →
+  [(Type, Name)] →
+  [BasicBlock] →
+  m Operand
+define retty label argtys body = do
+  addDefn
+    $ GlobalDefinition
+    $ functionDefaults
+      { Global.parameters = params,
+        -- Figure out which is best!
```

Comment:
Do we have to choose globally or can we choose per-function?

Reply:
Well the choices aren't much.

You can see them [here](http://hackage.haskell.org/package/llvm-hs-pure-9.0.0/docs/LLVM-AST.html#t:Definition).
I believe every function has to be global to the module. Note that we could always make function pointers and use that.

Reply:
Ah sorry, I should have commented on the line below, I meant the calling convention.